runhouse/rns/rns_client.py

```diff
@@ -220,18 +218,33 @@ def request_headers(
                 "Failed to extract token from request auth header. Expected in format: Bearer <token>"
             )
 
-        hashed_token = self.cluster_token(den_token, resource_address)
+        hashed_token = self.cluster_token(resource_address)
 
         return {"Authorization": f"Bearer {hashed_token}"}
 
-    def cluster_token(self, den_token: str, resource_address: str):
+    def cluster_token(self, resource_address: str):
+        """Load the hashed token as generated in Den."""
         if resource_address and "/" in resource_address:
             # If provided as a full rns address, extract the top level directory
             resource_address = self.base_folder(resource_address)
 
-        hash_input = (den_token + resource_address).encode("utf-8")
-        hash_hex = hashlib.sha256(hash_input).hexdigest()
-        return f"{hash_hex}+{resource_address}+{self._configs.username}"
+        uri = f"{self.api_server_url}/auth/token"
+        token_payload = {
+            "resource_address": resource_address,
+            "username": self._configs.username,
+        }
+        resp = self.session.post(
+            uri,
+            data=json.dumps(token_payload),
+            headers=self._configs.request_headers,
+        )
+        if resp.status_code != 200:
+            raise Exception(
+                f"Received [{resp.status_code}] from Den POST '{uri}': Failed to load cluster token: {load_resp_content(resp)}"
+            )
+
+        token_data = read_resp_data(resp)
+        return token_data.get("token")
```

Comment:
Should this method store the token in the config if it's not set already? To prevent future requests?

Reply:
I added a decorator to store the value of the cluster token for a given resource address an an in-mem cache for the life of the python process, otherwise i think could get messy if we're having to store lots of different tokens depending on the resource address. wdyt?

Reply:
Seems like a good solution to me! Is that cache invalidatible? Might need to do that if a token becomes stale but probably not an immediate problem